### PR TITLE
[1662] Allow missing keys

### PR DESCRIPTION
**Pull request by rijobro**
Fixes https://github.com/Project-MONAI/MONAI/issues/1662.

### Description
Option to not raise exception if keys are missing during transform.

This is my current proposition. I've only done it for a few classes to get opinions before continuing with rest.

The generator in the base class to minimise code duplication.

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/transforms/croppad/dictionary.py [hunk 1]
```diff
@@ -108,15 +109,16 @@ def __init__(
...
         """
-        super().__init__(keys)
+        super().__init__(keys, allow_missing_keys)
         self.mode = ensure_tuple_rep(mode, len(self.keys))
         self.padder = SpatialPad(spatial_size, method)
 
     def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.ndarray]:
         d = dict(data)
-        for key, m in zip(self.keys, self.mode):
+        for key, m in self.generator(d, self.mode):
```
Comment (rijobro) [inline monai/transforms/croppad/dictionary.py:121]: @wyli @Nic-Ma this is the approach if `zip` was previously used.

File: monai/transforms/transform.py [hunk 2]
```diff
@@ -224,3 +225,19 @@ def __call__(self, data):
 
         """
         raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    def generator(
```
Comment (wyli) [inline monai/transforms/transform.py:229]: perhaps we could have a more intuitive function name here...

**Review by wyli — COMMENTED** [review]
thanks, I think this a nice direction to go... @jansoe would be great to have your comments as well. I was also thinking of using a special key in `keys` such as `None` or `*` to represent the `allow_missing_keys` flag... but maybe that's more confusing APIs

**Review by jansoe — COMMENTED** [review]
Very cool, I think this will make pipelines much more flexible.

**Review by wyli — APPROVED** [review]
thank you, it looks good to me

**Pull request merged**